### PR TITLE
refactor: add schema to schema incompatibility error message

### LINKED ISSUE #3051: Better error when failing a statement due to schema incompatibility
When rejecting a query because of schema incompatibility, we should add the schema ksql wants to create in the response, and be explicit that the failure is due to an incompatibility. This way, the user can compare against the schema in the sr to see where the incompatibility id.

### PR DESCRIPTION
**Pull request by jzaralim**
### Description
Fixes #3051 
Added the schema in the error message for when an Avro schema is rejected by the schema registry.

 ## Testing done 
Updated unit tests.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java [hunk 1]
```diff
@@ -137,10 +138,15 @@ private void validateQuery(final QueryMetadata query, final ConfiguredStatement<
       final SchemaRegistryClient srClient = serviceContext.getSchemaRegistryClient();
 
       if (!AvroUtil.isValidSchemaEvolution(persistentQuery, srClient)) {
+        final org.apache.avro.Schema avroSchema = SchemaUtil.buildAvroSchema(
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java]: `AvroUtil.isValidSchemaEvolution` internally builds the avro schema.  So it feels like duplication to also build it here.

This is the only place `isValidSchemaEvolution` is called.  I think a slight refactor of your change to move the exception below into the method would be better, as it avoids the need to build the schema twice. I'd rename the function to `void throwOnInvalidSchemaEvolution(queryMd, srClient)`.

File: ksql-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java [hunk 2]
```diff
@@ -137,10 +138,15 @@ private void validateQuery(final QueryMetadata query, final ConfiguredStatement<
...
+            persistentQuery.getPhysicalSchema().valueSchema(),
+            persistentQuery.getSinkName()
+        );
         throw new KsqlStatementException(String.format(
-            "Cannot register avro schema for %s as the schema registry rejected it, "
-                + "(maybe schema evolution issues?)",
-            persistentQuery.getResultTopic().getKafkaTopicName()),
+            "Cannot register avro schema for %s as the schema is incompatible.%n"
+                + "schema: %s",
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java]: This exception is definitely better, as it includes the schema KSQL is trying to use.  What would be better still would be to include the schema in the schema registry in the message e.g.

`
Can not register avro schema for topic X as the schema is incompatible with the current schema version registered for the topic.
KSQL schema: {...}
Registered schema: {}
`

If you move the exception into the method, as discussed above, then in `testCompatibility` returns false you could grab the latest version of the schema using the following and include it in the exception as well:

```
// Compute this once:
final String subject = topicName + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX

// Grab latest schema using:
final String registeredSchema = schemaRegistryClient
          .getLatestSchemaMetadata(subject)
          .getSchema()
```

**Comment by jzaralim** [comment]
Thanks for the reviews @big-andy-coates! :)

File: ksql-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java [hunk 3]
```diff
@@ -115,7 +118,10 @@ ExecuteResult execute(final ConfiguredStatement<?> statement) {
     }
   }
 
-  private void validateQuery(final QueryMetadata query, final ConfiguredStatement<?> statement) {
+  private void validateQuery(
+      final QueryMetadata query,
+      final ConfiguredStatement<?> statement
+  ) throws IOException, RestClientException {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java]: We shouldn't be throwing checked exceptions here.  Any `IOException` should be caught internally and a better error message built.  (Though I don't think these exceptions are being thrown in this instance)

```suggestion
  ) {
```

File: ksql-engine/src/main/java/io/confluent/ksql/util/AvroUtil.java [hunk 4]
```diff
@@ -41,7 +42,27 @@ public static boolean isValidSchemaEvolution(
 
     final String topicName = queryMetadata.getResultTopic().getKafkaTopicName();
 
-    return isValidAvroSchemaForTopic(topicName, avroSchema, schemaRegistryClient);
+    if (!isValidAvroSchemaForTopic(topicName, avroSchema, schemaRegistryClient)) {
+      final String subject = topicName + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/AvroUtil.java]: Would you mind moving lines 45-54 into a method, e.g. `String getLatestSchema(topicName)`

File: ksql-engine/src/main/java/io/confluent/ksql/util/AvroUtil.java [hunk 5]
```diff
@@ -41,7 +42,27 @@ public static boolean isValidSchemaEvolution(
...
+      } catch (Exception e) {
+        registeredSchema = "Could not get registered schema due to exception: " + e.getMessage();
+      }
+
+      throw new KsqlStatementException(String.format(
+          "Cannot register avro schema for %s as the schema is incompatible with the current "
+              + "schema version registered for the topic.%n"
+              + "KSQL schema: %s%n"
+              + "Registered schema: %s",
+          queryMetadata.getResultTopic().getKafkaTopicName(),
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/AvroUtil.java]:
```suggestion
          topicName,
```

File: ksql-engine/src/main/java/io/confluent/ksql/util/AvroUtil.java [hunk 6]
```diff
@@ -41,7 +42,27 @@ public static boolean isValidSchemaEvolution(
...
+          avroSchema.toString(),
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/AvroUtil.java]:
```suggestion
          avroSchema,
```

File: ksql-engine/src/test/java/io/confluent/ksql/engine/KsqlEngineTest.java [hunk 7]
```diff
@@ -387,8 +387,20 @@ public void shouldFailIfAvroSchemaNotEvolvable() {
 
     expectedException.expect(KsqlStatementException.class);
     expectedException.expect(rawMessage(containsString(
-        "Cannot register avro schema for T as the schema registry rejected it, "
-            + "(maybe schema evolution issues?)")));
+        "Cannot register avro schema for T as the schema is incompatible with the current schema version registered for the topic.\n" +
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/engine/KsqlEngineTest.java]: Now that's a helpful error message!

**Pull request merged**